updated actix-session to support login and logout functionality

Dowwie opened the pull request:
To see the impact of these changes, see [main.rs in this repo](https://github.com/Dowwie/actix_session_test/blob/master/src/main.rs).

Specifically:
- L38:  invoke session.renew() at successful login
- L45:  session.purge() at logout

Dowwie committed:
updated actix-session to support login and logout functionality (renew and purge)

Dowwie committed:
modified so as to consider unanticipated state changes

fafhrd91 committed:
Merge branch 'master' into master

Dowwie committed:
made Session::get_session public

Dowwie committed:
Merge branch 'master' of https://github.com/Dowwie/actix-web

Dowwie commented:
Note that I currently cannot get a Session from within the scope of a ``from_request`` of an Extractor using UserSession trait because ``from_request`` receives an &HttpRequest where as a &mut HttpRequest is required for the UserSession trait's get_session to work.

In other words, I'm trying to get a Session from within an extractor and haven't found an alternative to cloning HttpRequest to do so.  One possible solution is to simply make actix-session/session.rs/SessionInner public and then I can replicate the logic in get_session to get it from extensions.  I don't need mutability in the case where Session already exists in request state and I *don't* want to create it if it it isn't.

Dowwie committed:
reverting change to get_session due to side effects

fafhrd91 commented:
you can always clone http request. also maybe get_session should take readonly ref

fafhrd91 committed:
Merge branch 'master' into master

Dowwie committed:
updated cookie.rs req to get_changes

Dowwie committed:
Merge branch 'master' into master

Dowwie commented:
@fafhrd91 looks good now?

Dowwie committed:
Merge branch 'master' into master

fafhrd91 commented on actix-session/src/cookie.rs:
```diff
@@ -308,7 +308,7 @@ where
 
         Box::new(self.service.call(req).map(move |mut res| {
-            if let Some(state) = Session::get_changes(&mut res) {
+            if let (_status, Some(state)) = Session::get_changes(&mut res) {
```
we should update cookies only if session has changes, otherwise we should not set cookie. i think this is the reason for #923

Dowwie replied:
the latest commit should reflect this

fafhrd91 reviewed (changes requested):
looks good, except cookie session

Dowwie committed:
updated cookie session to update on change

Dowwie commented:
> looks good, except cookie session

@fafhrd91 I added a commit that I believe is what you were asking for -- confirm?  https://github.com/actix/actix-web/pull/912/commits/0e05b37082fa57444858b46e1a8b57b731777f1c

@max-frai -- Can you run this PR through your integration tests and confirm that this change addresses your issues in [923](https://github.com/actix/actix-web/issues/923)?  I'm not using cookie sessions but rather redis sessions so I haven't got integration tests for this side.

Dowwie commented:
One important note about this new approach is that when a programmer wants to create a session with the first request to a web server , the only way to create the session is to set state.  Currently, a new, empty session is created upon first request.  The way that the current session logic accomplishes the session creation is to create a session entry with empty state -- no changes.  I'm leaning towards replicating this to keep behavior consistent.  It's not hard work to add support for initializing a session upon first request, even if its empty and has no changes.  

Update:  I've addressed this in a subsequent commit

fafhrd91 commented on actix-session/src/cookie.rs:
```diff
@@ -308,10 +308,10 @@ where
-            } else {
-                res
+            match Session::get_changes(&mut res) {
+                (SessionStatus::Changed, Some(state)) =>
```
i think we should update cookie on `change` and `renew`, and delete cookie on `purge`

Dowwie replied:
done

fafhrd91 commented:
i am fine with new state. i added comment for cookie session

Dowwie committed:
updated session cookie to support login, logout, changes

Dowwie commented:
@fafhrd91 the latest commit addresses unchanged update to cookie and adds support for purging cookie -- should be ready.

fafhrd91 commented:
@Dowwie could you also update changes.md for actix-session

Dowwie committed:
updated CHANGES with info about actix-session update

Dowwie committed:
updated actix-session/CHANGES with info

Dowwie committed:
updated version in Cargo to 0.2

Dowwie commented:
> @Dowwie could you also update changes.md for actix-session

I updated actix-session CHANGES and actix-web CHANGES.  This release introduces SessionState and new Session behaviors -- renew and purge.  I thought this is sufficient to bump actix-session from 0.1.x to 0.2 and so updated Cargo.toml and CHANGES respectively.

Dowwie committed:
reworded session info in CHANGES

Dowwie committed:
reverted actix-web/CHANGES.md

Dowwie committed:
Merge branch 'master' into master
